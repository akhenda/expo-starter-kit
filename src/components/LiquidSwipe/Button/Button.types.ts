import type { ViewProps } from 'react-native';
import type Animated from 'react-native-reanimated';
import type { Vector } from 'react-native-redash';

import type { Side } from '@components/LiquidSwipe/Wave';

export interface ButtonProps extends ViewProps {
  position: Vector<Animated.SharedValue<number>>;
  side: Side;
  activeSide: Animated.SharedValue<Side>;
}
