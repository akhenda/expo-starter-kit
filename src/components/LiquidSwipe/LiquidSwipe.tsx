import { useState } from 'react';

import { images } from '@src/theme';

import { Slide, SlideProps } from './Slide';
import { Slider } from './Slider';

const slides: SlideProps['slide'][] = [
  {
    color: '#F2A1AD',
    description: 'Hot or cold, our dessert recipes can turn an average meal into a memorable event',
    picture: images.onboarding.liquidSwipe[0],
    title: 'Dessert Recipes',
  },
  {
    color: '#0090D6',
    description: 'Discover healthy recipes that are easy to do with detailed cooking instructions from top chefs',
    picture: images.onboarding.liquidSwipe[4],
    title: 'Healthy Foods',
  },
  {
    color: '#69C743',
    description: 'explore recipes by food type, preparation method, cuisine, country and more',
    picture: images.onboarding.liquidSwipe[3],
    title: 'Easy Meal Ideas',
  },
  {
    color: '#FB3A4D',
    description: 'Browse thousands of curated recipes from top chefs, each with detailled cooking instructions',
    picture: images.onboarding.liquidSwipe[1],
    title: '10000+ Recipes',
  },
  {
    color: '#F2AD62',
    description: 'Browse our best themed recipes, cooking tips, and how-to food video & photos',
    picture: images.onboarding.liquidSwipe[2],
    title: 'Video Tutorials',
  },
];

export const assets = slides.map(({ picture }) => picture);

const LiquidSwipe = () => {
  const [index, setIndex] = useState(0);

  const prev = slides[index - 1];
  const next = slides[index + 1];

  return (
    <Slider
      key={index}
      index={index}
      setIndex={setIndex}
      prev={prev && <Slide slide={prev} />}
      next={next && <Slide slide={next} />}
    >
      <Slide slide={slides[index] as SlideProps['slide']} />
    </Slider>
  );
};

export default LiquidSwipe;
