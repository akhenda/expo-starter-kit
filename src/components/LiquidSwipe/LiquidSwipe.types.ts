import type { StyleProp, ViewProps, ViewStyle } from 'react-native';

export interface LiquidSwipeProps extends ViewProps {
  /**
   * Extra styles for the container
   */
  style?: StyleProp<ViewStyle>;
}
