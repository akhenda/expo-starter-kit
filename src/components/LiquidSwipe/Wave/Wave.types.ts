import type { ReactElement } from 'react';
import type { ViewProps } from 'react-native';
import type Animated from 'react-native-reanimated';
import type { Vector } from 'react-native-redash';

import type { SlideProps } from '@components/LiquidSwipe/Slide';

export enum Side {
  LEFT,
  RIGHT,
  NONE,
}

export interface WaveProps extends ViewProps {
  side: Side;
  position: Vector<Animated.SharedValue<number>>;
  children: ReactElement<SlideProps>;
  isTransitioning: Animated.SharedValue<boolean>;
}
