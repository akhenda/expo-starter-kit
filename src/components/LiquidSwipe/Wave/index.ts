export * from './Wave';
export { default as Wave } from './Wave';
export * from './Wave.types';
