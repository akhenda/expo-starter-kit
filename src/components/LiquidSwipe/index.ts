export { default as LiquidSwipe } from './LiquidSwipe';
export * from './LiquidSwipe.types';
