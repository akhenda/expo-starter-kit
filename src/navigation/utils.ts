import React, { useCallback, useEffect, useLayoutEffect, useRef, useState } from 'react';
import { BackHandler } from 'react-native';
import { createNavigationContainerRef, NavigationState, useNavigation } from '@react-navigation/native';

import type { PersistNavigationConfig } from '@config/config.base';
import Config from '@src/config';
import { useIsMounted } from '@src/hooks/useLifecycle';
import { mmkvStorage as storage } from '@src/services/storage';

import type { RootStackParamList } from './navigators/RootNavigator';

export const NAVIGATION_PERSISTENCE_KEY = '@NAVIGATION_STATE';

export const navigationRef = createNavigationContainerRef<RootStackParamList>();

/**
 * Gets the current screen from any navigation state.
 */
export function getActiveRouteName(state?: NavigationState<RootStackParamList>): keyof RootStackParamList {
  if (!state || !state.index) return 'App Stack';

  const route = state.routes[state.index];

  if (!route) return 'App Stack';

  // Found the active route -- return the name
  if (!route.state) return route.name;

  // Recursive call to deal with nested routers
  return getActiveRouteName(route.state as NavigationState<RootStackParamList>);
}

/**
 * Hook that handles Android back button presses and forwards those on to
 * the navigation or allows exiting the app.
 */
export function useBackButtonHandler(canExit: (routeName: keyof RootStackParamList) => boolean) {
  // The reason we're using a ref here is because we need to be able
  // to update the canExit function without re-setting up all the listeners
  const canExitRef = useRef(canExit);

  useEffect(() => {
    canExitRef.current = canExit;
  }, [canExit]);

  useEffect(() => {
    // We'll fire this when the back button is pressed on Android.
    const onBackPress = () => {
      if (!navigationRef.isReady()) return false;

      // grab the current route
      const routeName = getActiveRouteName(navigationRef.getRootState() as NavigationState<RootStackParamList>);

      // are we allowed to exit?
      if (canExitRef.current(routeName)) {
        // exit and let the system know we've handled the event
        BackHandler.exitApp();

        return true;
      }

      // we can't exit, so let's turn this into a back action
      if (navigationRef.canGoBack()) {
        navigationRef.goBack();

        return true;
      }

      return false;
    };

    // Subscribe when we come to life
    BackHandler.addEventListener('hardwareBackPress', onBackPress);

    // Unsubscribe when we're done
    return () => BackHandler.removeEventListener('hardwareBackPress', onBackPress);
  }, []);
}

/**
 * This helper function will determine whether we should enable navigation persistence
 * based on a config setting and the __DEV__ environment (dev or prod).
 */
function navigationRestoredDefaultState(persistNavigation: PersistNavigationConfig) {
  if (persistNavigation === 'always') return false;
  if (persistNavigation === 'dev' && __DEV__) return false;
  if (persistNavigation === 'prod' && !__DEV__) return false;

  // all other cases, disable restoration by returning true
  return true;
}

/**
 * Custom hook for persisting navigation state.
 */
export function useNavigationPersistence(persistenceKey: string = NAVIGATION_PERSISTENCE_KEY) {
  const isMounted = useIsMounted();
  const routeNameRef = useRef<string | undefined>();
  const initNavState = navigationRestoredDefaultState(Config.persistNavigation);

  const [initialNavigationState, setInitialNavigationState] = useState<NavigationState<RootStackParamList>>();
  const [isRestored, setIsRestored] = useState(initNavState);

  const onNavigationStateChange = (state?: NavigationState) => {
    const previousRouteName = routeNameRef.current;
    const currentRouteName = getActiveRouteName(state as NavigationState<RootStackParamList>);

    if (previousRouteName !== currentRouteName) {
      // track screens.
      if (__DEV__)
        // eslint-disable-next-line no-console
        console.tron.display?.({
          name: 'currentRouteName',
          preview: currentRouteName,
          value: currentRouteName,
        });
    }

    // Save the current route name for later comparison
    routeNameRef.current = currentRouteName;

    // Persist state to storage
    storage.save(persistenceKey, state);
  };

  const restoreState = useCallback(async () => {
    try {
      const state = (await storage.load(persistenceKey)) as NavigationState<RootStackParamList>;

      if (state) setInitialNavigationState(state);
    } finally {
      if (isMounted()) setIsRestored(true);
    }
  }, [persistenceKey, isMounted]);

  useEffect(() => {
    if (!isRestored) restoreState();
  }, [isRestored, restoreState]);

  return { initialNavigationState, isRestored, onNavigationStateChange, restoreState };
}

/**
 * A hook that can be used to easily set the Header of a react-navigation screen from within the screen's component.
 *
 * - [Documentation and Examples](https://github.com/infinitered/ignite/blob/master/docs/Utils-useHeader.md)
 */
export function useHeader(header: React.ReactNode, deps: unknown[] = []) {
  const navigation = useNavigation();

  useLayoutEffect(() => {
    navigation.setOptions({
      header,
      headerShown: true,
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [...deps, header, navigation]);
}

/**
 * use this to navigate without the navigation
 * prop. If you have access to the navigation prop, do not use this.
 * More info: https://reactnavigation.org/docs/navigating-without-navigation-prop/
 */
export function navigate(...args: Parameters<typeof navigationRef.navigate>): void {
  if (navigationRef.isReady()) navigationRef.navigate(...args);
}

export function goBack() {
  if (navigationRef.isReady() && navigationRef.canGoBack()) navigationRef.goBack();
}

export function resetRoot(params = { index: 0, routes: [] }) {
  if (navigationRef.isReady()) navigationRef.resetRoot(params);
}
