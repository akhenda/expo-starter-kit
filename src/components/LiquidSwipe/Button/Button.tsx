import Animated, { useAnimatedStyle, withTiming } from 'react-native-reanimated';
import { Feather as Icon } from '@expo/vector-icons';

import { Side } from '@components/LiquidSwipe/Wave';
import metrics from '@src/theme/metrics';

import type { ButtonProps } from './Button.types';

const RADIUS = 25;
const { width } = metrics;

const Button = ({ position, side, activeSide }: ButtonProps) => {
  const isLeft = side === Side.LEFT;
  const style = useAnimatedStyle(() => ({
    alignItems: 'center',
    borderRadius: RADIUS,
    height: RADIUS * 2,
    justifyContent: 'center',
    left: isLeft ? position.x.value - RADIUS * 2 : width - position.x.value,
    opacity: withTiming(activeSide.value === Side.NONE ? 1 : 0),
    position: 'absolute',
    top: position.y.value - RADIUS,
    width: RADIUS * 2,
  }));

  return (
    <Animated.View style={style}>
      <Icon name={`chevron-${isLeft ? 'right' : 'left'}` as const} size={24} color="white" />
    </Animated.View>
  );
};

export default Button;
