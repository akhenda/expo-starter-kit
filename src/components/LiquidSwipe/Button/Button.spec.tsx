import type { SharedValue } from 'react-native-reanimated';
import { cleanup, screen } from '@testing-library/react-native';

import { Side } from '@components/LiquidSwipe/Wave';
import { render } from '@src/utils/tests';

import Button from './Button';

afterEach(cleanup);

it('renders correctly', async () => {
  render(
    <Button
      position={{ x: 0 as unknown as SharedValue<number>, y: 0 as unknown as SharedValue<number> }}
      side={Side.LEFT}
      activeSide={Side.LEFT as unknown as SharedValue<Side>}
    />,
  );

  expect(screen.toJSON()).toMatchSnapshot();
});
