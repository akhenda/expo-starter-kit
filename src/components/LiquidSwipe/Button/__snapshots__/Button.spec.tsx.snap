// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "borderRadius": 25,
      "height": 50,
      "justifyContent": "center",
      "left": NaN,
      "opacity": 0,
      "position": "absolute",
      "top": NaN,
      "width": 50,
    }
  }
>
  <Text />
</View>
`;
